### DISCUSSION TITLE
Ability to set and refresh token in HEADERS visiting pages by inertia-link

### DISCUSSION THREAD

**Discussion by zablik** [inertiajs/inertia#454] [Help] [answered]
Hi everyone!

I'm building an application where authentication is done by OpenID (oidc-client-js). Tokens are issued by the Authentication center (it's not my app)
My SPA on Inertia.js obtains a token and sends it to my backend part, where I validate token's signature.
As the token is short-living I need the ability to refresh it and restrict access to any controller since the token is expired.

I know about manual visits `Inertia.visit()` but navigation should be done with `<inertia-links />`, doesn't it?
Also, I know about `:headers` attribute.

Looks like this approach works. But it does not look pretty...
```
window.axios = require("axios");
// and then periodically call
window.axios.defaults.headers.common["x-auth"] = "my token"
```

So, my question is: what way do you recommend to refresh headers while navigating by inertia-links?

Thanx! =)

**Comment by ajnsn** [comment, marked as answer]
@zablik In a typical Inertia/Laravel app there's some magic behind the scenes as the `XSRF-TOKEN` cookie is automatically added by Axios to every request and refreshed by Laravel when returning a 419. You just need redirect back and show a message to the user, like [described here](https://inertiajs.com/csrf-protection#handling-mismatches). For`GET` requests there are typically no tokens needed at all.

In your case, things seem different. Here are some existing issues/answers, you could take a look at.
- You could hook into the events of Inertia ([Make `headers` available in `start` event](https://github.com/inertiajs/inertia/issues/242#issuecomment-702274376))
- or try to use Axios interceptors ([How to add custom headers?](https://github.com/inertiajs/inertia/issues/147#issuecomment-655922157)). You just make sure you are on the same version as Inertia, see [Opportunity for better 419 handling](https://github.com/inertiajs/inertia/issues/367#issuecomment-751353535).

Also note that there are plans to remove Axios completely. The Axios approach could not be too persistent.

**Reply by zablik** [reply]
Thanks, I think I need some time for investigation. 
And I hope with removing Axios an alternative way of working with headers like in my case will be provided =)
